sklearn/utils/_indexing.py

```diff
@@ -560,11 +566,21 @@ def resample(*arrays, replace=True, n_samples=None, random_state=None, stratify=
     # convert sparse matrices to CSR for row-based indexing
     arrays = [a.tocsr() if issparse(a) else a for a in arrays]
     resampled_arrays = [_safe_indexing(a, indices) for a in arrays]
+
+    if sample_weight is not None:
+        resampled_weights = _safe_indexing(sample_weight, indices)
```

Comment:
Good question (I'm also not sure), we could do that but in principle the KBinsDiscretizer takes sample weight into account when calculating the quantile (using _weighted_quantile) or KMeans based bins. So if we resample using weights we would have to remove those parts from KBinsDiscretiser. We would also then have to check other estimators using this (e.g., HistGradientBoostingRegressor). 

I did some preliminary checks on the sample_weight_invariance using the approach of weighted resampling and moving further use of weights once we sample the imbalanced dataset but they didn't perform so well... not sure what you think would be a correct approach?

Reply:
From my limited experience working on `KBinsDiscretizer` and my understanding of the transformation I would say uniform sampling and weighted quantile is the right way to do it.

Reply:
> Maybe there is a way to check which one is better?

Let's check both options in a notebook and look at the distribution of the bin edges via histograms when calling it many times with different random seeds and checking if one has a better match with the equivalent repeated dataset (assuming integer weights). The match can be checked in 3 different ways:

- qualitatively via histogram plots;
- via a KS-test to quantify the match in distribution;
- via a Welch's t-test to quantify the match in expected value.

Reply:
> * via a Welch's t-test to quantify the match in expected value.

Great so I already did the Welch t-test on the bin edges with results attached below:
<img width="1163" alt="Screenshot 2024-07-04 at 16 00 04" src="https://github.com/user-attachments/assets/aefe035e-2eed-4eae-bbdc-b5f534e6aeca">

I check both kmeans and quantile strategy with (1) passing the weights to resampling (Note here I make sure it is resampling with replacement) and (2) passing them to the quantile/kmeans calculation. The results seem a bit mixed and strategy dependent (I show the t-test minimum p-value and average p-value over 5 representative samples, results are sorted by minimum p-value)

> 
> * qualitatively via histogram plots;
> * via a KS-test to quantify the match in distribution;

I will work on these, hopefully they make the picture a bit clearer

Reply:
After looking at the code, I see that there is no weighted resampling. It just uses `numpy.random.randint`, `numpy.random.shuffle` or `numpy.random.choice` without any account for sample weights. The only thing it does is extracting the sample weights that correspond to samples drawn and returning them at the end.
Basically it is either rewrite `resample` function so it would use `numpy.random.choice` everywhere with `sample_weight` passed to `p` parameter, or just use uniform sampling with weighted percentiles.

Reply:
Yes, so the table above compares the two different paradigms of passing weights to the KBinsDiscretizer:
1) Where you use numpy.random.choice in the resample function 
2) Where we uniformly sample but use weighted percentiles/kmeans
If you look at the first column you can see which p-values corresponsd to which strategy: ideally we do not want significant p-values (i.e., all should be above 0.025, meaning that different random states lead to the same predicted bin edges in expectation)

Reply:
So I tried to plot the distribution of bin_edges that we get with random states 0...1 with subsampling allowed. The histograms for when I use uniform resampling with the weighted quantile is as follows:

![image](https://github.com/user-attachments/assets/15fd0758-02e3-4466-8acb-6bbfb24c0e6e)

Reply:
A similar plot but for when I use weighted resampling with replacement and the normal quantile bin calculation is as follows:

![image](https://github.com/user-attachments/assets/77994455-f950-4c39-a24f-8cc14e464e4b)

Reply:
The average kstest score between the two methods are both almost 0.154.

When I try the same with the kmeans strategy instead I however get an error for the uniform resampling with weighted kmeans scenario:

```
ConvergenceWarning: Number of distinct clusters (4) found smaller than n_clusters (5). Possibly due to duplicate points in X.
  return fit_method(estimator, *args, **kwargs)
```

This is not the case for when I do weighted resampling with replacement that is then passed on to the KMeans, which provides the following histograms for the bin edges:

![image](https://github.com/user-attachments/assets/ddd8dca8-8313-42bb-982a-b9dffff3f550)

and average ks-test value of 0.301

Reply:
@jeremiedbb any idea why the KMeans convergence warning occurs?

Reply:
What do you mean by "ks-test value"? Is this the p-value (a lower value means rejecting the null-hypothesis that the distributions match?)? Or the value of the ks-test statistic (the supremum of the absolute difference of the estimated CDF for the two distributions)? If the latter, a large value means a lack of match between distributions.

Maybe you could link to the full notebook of this analysis of the distribution for the bin edges? This would remove any ambiguity. Also, maybe increasing the number of seeds used to run the simulation would make both the qualitative histogram conclusions and the quantitative KS-test results easier to interpret.

To be able to test all strategies in the same notebook, feel free to reimplement a minimal version of the edge computation directly in the notebook itself, instead of relying on various branches of the scikit-learn code base.

Reply:
BTW, the warning "UserWarning: Bins whose width are too small (i.e., <= 1e-8) in feature 1 are removed. Consider decreasing the number of bins." is not necessarily a big problem for actual use of the method, but it would indeed make it harder to compare different methods to compute bin-edges if the resulting number of bins is not always fixed.

So maybe re-run this analysis with a smaller number of bins (e.g. 5 instead of 10). This should reduce the likelihood to getting too small differences between consecutive bin edges.

Reply:
Also, please systematically report both the p-value and the KS-test statistic value for each experiment, along with the bin edge histograms.

Reply:
> What do you mean by "ks-test value"? Is this the p-value (a lower value means rejecting the null-hypothesis that the distributions match?)? Or the value of the ks-test statistic (the supremum of the absolute difference of the estimated CDF for the two distributions)? If the latter, a large value means a lack of match between distributions.
> 
It's the ks-test p-value so smaller values is a lack of match, I will prepare the notebook and link it then :)

Reply:
I'm not sure KS test is a good test here. Since we will compare discrete distributions there will be ties and it will decrease the power of KS test. T-tets or maybe something like chi2 should be better.